### PR TITLE
Add Preview Image to Card Creation and Frontend Error Handling

### PR DESCRIPTION
**Pull request by ESaracay**
Screenshot 2020-08-12 at 3 33 35 PM - Display 2
This PR adds previews to the card creation page for both images and text translation.

### CODE REVIEW

File: frontend/src/main-components/CreateCardContent.js [hunk 1]
```diff
@@ -15,27 +15,57 @@ class CreateCardContent extends Component {
...
     this.fromLangSelected = this.fromLangSelected.bind(this);
     this.toLangSelected = this.toLangSelected.bind(this);
+    this.imageSelected = this.imageSelected.bind(this);
+    this.imageUploadUrl = "";
+  }
+
+  async componentDidMount() {
+    try {
+      const uploadUrl = await fetch("/upload")
+        .then((response) => { return response.text() });
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: Nit: you can write this as just

```js
.then(response => response.text());
```

File: frontend/src/main-components/CreateCardContent.js [hunk 2]
```diff
@@ -15,27 +15,57 @@ class CreateCardContent extends Component {
...
+      // Throws error if the /upload servlet is not found
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: necessary?

File: frontend/src/main-components/CreateCardContent.js [hunk 3]
```diff
@@ -15,27 +15,57 @@ class CreateCardContent extends Component {
...
+    }
   }
 
   /**
    * When the user has finished typing, the Google Translate
    * API is called to fetch the translated version of the text input.
    */
-  translateText(event) {
+  async translateText(event) {
+    event.persist();
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: Why do we need to persist the event?

File: frontend/src/main-components/CreateCardContent.js [hunk 4]
```diff
@@ -15,27 +15,57 @@ class CreateCardContent extends Component {
...
+    const target = event.target.value;
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: Is target the best name for this variable?

File: frontend/src/main-components/CreateCardContent.js [hunk 5]
```diff
@@ -15,27 +15,57 @@ class CreateCardContent extends Component {
...
+          this.state.fromLang,
+          this.state.toLang
+        );
+        // Checks if the getTranslation function encountered an error
+        if (!translated.ok) {
+          throw Error(translated.statusText)
+        }
+        this.setState({ translation: translated, text: target });
+      } catch (error) {
+          this.setState({ translation: "Could not Translate", text: target });
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: Grammar?

File: frontend/src/main-components/CreateCardContent.js [hunk 6]
```diff
@@ -15,27 +15,57 @@ class CreateCardContent extends Component {
...
+      }
+    } else {
+        this.setState({ text: target});
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: `target })`

**Review by tylerhou — CHANGES_REQUESTED** [review]

File: frontend/src/main-components/CreateCardContent.js (hunk 3 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: It has to be used in async functions if you want to get the value from an event. In this case were getting the text from one of the input fields.

File: frontend/src/main-components/CreateCardContent.js (hunk 6 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js (hunk 5 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js (hunk 4 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js (hunk 3 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: Replaced with immediately invoked async function.

Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js (hunk 2 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js (hunk 1 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js [hunk 7]
```diff
@@ -15,27 +15,60 @@ class CreateCardContent extends Component {
...
-      );
-      this.setState({ translation: translated, text: event.target.value });
+      (async () => {
+        try {
+          const translated = await getTranslation(
+            text,
+            this.state.fromLang,
+            this.state.toLang
+          );
+          // Checks if the getTranslation function encountered an error
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: Maybe don't need this comment?

**Review by tylerhou — APPROVED** [review]

File: frontend/src/main-components/CreateCardContent.js [hunk 8]
```diff
@@ -15,27 +15,60 @@ class CreateCardContent extends Component {
...
+          this.setState({ translation: translated, text: text });
+        } catch (error) {
+          this.setState({
+            translation: "Text could not be translated",
+            text: text,
+          });
+        }
+      })();
+    } else {
+      this.setState({ text: text });
```
Comment (tylerhou) [inline frontend/src/main-components/CreateCardContent.js]: Small nit: when you have a variable with the same name as the key in the object, there's a shortcut:

```js
      this.setState({ text });
```

Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: This is a cool tip, thank you.

Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

File: frontend/src/main-components/CreateCardContent.js (hunk 7 above)
Comment (ESaracay) [reply frontend/src/main-components/CreateCardContent.js]: done

**Pull request merged**